### PR TITLE
Direct Access to parse-server

### PR DESCRIPTION
**Pull request by flovilmart**
This PR enables direct access to parse-server, when using the JS SDK from the current node runtime.

All API calls from the JS-SDK are router through route matching (but /functions/:functionName)

Enable that feature by setting PARSE_SERVER_ENABLE_EXPERIMENTAL_DIRECT_ACCESS=1

### CODE REVIEW

**Comment by facebook-github-bot-6** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-6** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-2** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@flovilmart updated the pull request.

File: spec/PushController.spec.js [hunk 1]
```diff
@@ -347,13 +347,15 @@ describe('PushController', () => {
...
-     });
+     setTimeout(() => {
+      let query = new Parse.Query('_PushStatus');
+      query.find({useMasterKey: true}).then((results) => {
+        expect(results.length).toBe(1);
+        let pushStatus = results[0];
+        expect(pushStatus.get('status')).toBe('failed');
+        done();
+      });
+     }, 500);
```
Comment (flovilmart) [inline spec/PushController.spec.js]: Need to add some delay here as with DIRECT_ACCESS the status is not updated upon failure...

**Comment by facebook-github-bot-2** [comment]
@flovilmart updated the pull request.

**Comment by flovilmart** [comment]
Closing for now as will rebase when #2346 is merged

**Comment by andrewimm** [comment]
This sounds like a great approach. I'll be sure to review when you reopen.

**Comment by flovilmart** [comment]
If you have time to review #2346 I can get going, I'd like to have that merge pretty soon

**Comment by facebook-github-bot-2** [comment]
@flovilmart updated the pull request.

**Comment by flovilmart** [comment]
@andrewimm I just reopened it, and tweaked a few parts to make it nicer.

**Comment by facebook-github-bot-2** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-2** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-2** [comment]
It's been a while since the last commit was reviewed and the labels show this pull request needs review. Based on the blame information for the files in this pull request we identified @acinader as a potential reviewer. Could you take a look please or cc someone with more context?

File: spec/CloudCodeLogger.spec.js [hunk 2]
```diff
@@ -3,55 +3,55 @@ var LoggerController = require('../src/Controllers/LoggerController').LoggerCont
 var FileLoggerAdapter = require('../src/Adapters/Logger/FileLoggerAdapter').FileLoggerAdapter;
 
 describe("Cloud Code Logger", () => {
```
Comment (acinader) [inline spec/CloudCodeLogger.spec.js]: just outta curiosity, why the change here?  it looks to me like you are just swapping the order of two tests?  or am i missing something else...

File: src/ParseServer.js [hunk 3]
```diff
@@ -56,6 +56,7 @@ import DatabaseController       from './Controllers/DatabaseController';
 const SchemaController = require('./Controllers/SchemaController');
 import ParsePushAdapter         from 'parse-server-push-adapter';
 import MongoStorageAdapter      from './Adapters/Storage/Mongo/MongoStorageAdapter';
+import { DirectRESTController } from './DirectRESTController';
```
Comment (acinader) [inline src/ParseServer.js]: just my js ignoreance, but why is this import {} while the others aren't?

File: src/ParseServer.js [hunk 4]
```diff
@@ -322,6 +323,10 @@ class ParseServer {
 
     api.use(middlewares.handleParseErrors);
 
+    if (process.env.PARSE_SERVER_ENABLE_EXPERIMENTAL_DIRECT_ACCESS == '1') {
```
Comment (acinader) [inline src/ParseServer.js]: === :)

**Comment by acinader** [comment]
The bot asked me to look, so I'm looking.  Not really adding any value at pr's (yet), but good to read the code.

My big question on this one is "why?".  Would be helpful to me, though not necessarily helpful to the code review, if you could give me a use case to understand what is going on.

File: src/DirectRESTController.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+          }
+        })
+      }
+
+      let query;
+      if (method === 'GET') {
+        query = data;
+      }
+
+      logRequest("internal"+path, method, data, {});
```
Comment (acinader) [inline src/DirectRESTController.js]: add ' ' around +, or better yet, use template

File: src/DirectRESTController.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            query
+          };
+          return Promise.resolve().then(() => {
+            return router.tryRouteRequest(method, path, request);
+          }).then((response) => {
+              logResponse("internal"+path, method, response.response);
+              resolve(response.response, response.status, response);
+          }, (err) => {
+            if (err instanceof Parse.Error &&
+                err.code == Parse.Error.INVALID_JSON &&
```
Comment (acinader) [inline src/DirectRESTController.js]: ===

File: src/batch.js [hunk 7]
```diff
@@ -1,5 +1,5 @@
 var Parse = require('parse/node').Parse;
-
+var routeRESTRequest = require('./DirectRESTController').routeRESTRequest;
```
Comment (acinader) [inline src/batch.js]: why var not let when you use let a few lines down?

**Comment by acinader** [comment]
ok, i reviewed it, but since i don't really understand what is going on, its only of value to me, not you ;).

File: src/ParseServer.js (hunk 3 above)
Comment (flovilmart) [inline src/ParseServer.js]: When not using {} this is to import the `default`, with {} it's the spread operator to get the properties of the exported object.

File: src/batch.js (hunk 7 above)
Comment (flovilmart) [inline src/batch.js]: :+1:

File: src/ParseServer.js (hunk 3 above)
Comment (acinader) [inline src/ParseServer.js]: wow, that actually makes sense!  thanks.

**Comment by zleman1593** [comment]
@flovilmart  Are you planning on merging this soon?  Looking forward to it.

**Comment by flovilmart** [comment]
Not soon, it needs some rebase as well as fine tuning for cloud code usage. Mostly a custom UserController for the Parse SDK.

File: src/DirectRESTController.js [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        return Parse.Promise.all(promises);
+      }
+
+      let config = new Config(applicationId);
+      let serverURL = URL.parse(config.serverURL);
+      if (path.indexOf(serverURL.path) === 0) {
+        path = path.slice(serverURL.path.length, path.length);
+      }
+
+      if (path[0] !== "/") {
```
Comment (acinader) [inline src/DirectRESTController.js]: ' instead of "

**Comment by acinader** [comment]
ok, i get it now.  cool.  you want me to test on our dev servers, or should i want for the cloud code changes?

**Comment by flovilmart** [comment]
Yeah needs rebase and fix for that crappy UserController.

**Comment by facebook-github-bot-4** [comment]
@flovilmart updated the pull request - [view changes](https://github.com/ParsePlatform/parse-server/pull/2316/files/3eafa048b0b6ceffb10f312b57b83247f50f2713..2309f8b8fba8ed5c452aa313f8cab7b1931180cf)

**Comment by facebook-github-bot-5** [comment]
@flovilmart updated the pull request - [view changes](https://github.com/ParsePlatform/parse-server/pull/2316/files/2309f8b8fba8ed5c452aa313f8cab7b1931180cf..ce2f1bd4984ce7acc7a49c833766153bd0a21618)

**Comment by flovilmart** [comment]
@acinader ready to test in your environment. I could very well make an 'alpha tag' on NPM so it's easy to install :)

**Comment by acinader** [comment]
@flovilmart the npm tag would be useful since I am going to deploy it to dev servers and not just work locally.  

I can deploy, but man, its some serious gymnastics.

I probably wont get to until this weekend.

**Comment by flovilmart** [comment]
Awesome! I'll try to do that :)

**Comment by flovilmart** [comment]
@acinader the alpha is released, you can use it :) let me know your finding. There won't be no verbose logs on the direct requests, I could add them if you think they should be included.

**Comment by flovilmart** [comment]
@andrewimm if you got a minute to check that out again, see if there is anything that could be improved here.

**Comment by facebook-github-bot-4** [comment]
@flovilmart updated the pull request - [view changes](https://github.com/ParsePlatform/parse-server/pull/2316/files/ce2f1bd4984ce7acc7a49c833766153bd0a21618..b5cc3146bfc306f0134efb8e1a99501ca11d47ef)

**Comment by facebook-github-bot-9** [comment]
@flovilmart updated the pull request - [view changes](https://github.com/ParsePlatform/parse-server/pull/2316/files/b5cc3146bfc306f0134efb8e1a99501ca11d47ef..e3098fc1189f644d6cd01119d39d5fffb702a263)

**Comment by flovilmart** [comment]
Provide a fix for #1495

When using the DirectRestController, the sessionToken won't be created when saving a new user.

**Comment by acinader** [comment]
@flovilmart I did get around to getting set up and running the alpha tag, but I haven't had a chance to run through some of the account creation, editing, cloud code, tests that I want to do.  Another day or two. But no issues in my very casual testing this weekend.

**Comment by flovilmart** [comment]
Alright that's good to hear! As long as you don't use Parse.User.currentUser() and Parse.Cloud.useMasterKey() that should be fine.

On another note, the installationId for the requests are 'cloud' so we don't generate sessionTokens. I believe it should have very little effet on your code.

**Comment by flovilmart** [comment]
Alright! Keep me posted!

**Comment by flovilmart** [comment]
@acinader any update on your side?

**Comment by drew-gross** [comment]
If this passes tests I'd be fine with skipping the experimental flag stage. We have a pretty comprehensive test suite, it would be pretty tough to make an implementation that passes tests but still has major issues.

**Comment by acinader** [comment]
👍 we ran two releases through qa with alpha on our dev boxes.  one with direct on and one with it off.  I wanted to do ~~better~~ any profiling to see impact, but that's still gotta wait.

@flovilmart sorry for delay.

**Comment by flovilmart** [comment]
It's gonna be in the next release, 2.2.19, still marked as experimental until we have a better idea on how to make sure it works :100:

So you'll be able to go back to the regular release cycle with the experimental flag on

**Comment by listingboat** [comment]
@flovilmart So, this is in the current version as experimental.  I turned on PARSE_SERVER_ENABLE_EXPERIMENTAL_DIRECT_ACCESS on our Heroku instance and saw faster performance, but also some issues.  Our app is a cloud code web app hosted on Heroku.

I noticed in above comments:
"As long as you don't use Parse.User.currentUser() and Parse.Cloud.useMasterKey() that should be fine."

Why is that?  We do use Parse.Cloud.useMasterKey to update Users.  I can dig further if needed.

**Comment by flovilmart** [comment]
you should not use Parse.Cloud.useMasterKey() at all.

Instead use:

user.save(null, {useMasterKey: true})

**Comment by nitrag** [comment]
What's the status on this feature in terms of it becoming the standard? I don't see any recent Issues. Is it safe to say it's production ready?

**Comment by flovilmart** [comment]
We need to revamp some tests but yes, we can probably make it mainline.

**Comment by mtrezza** [comment]
This looks like a PR that would give a significant performance improvement without any breaking changes.

@acinader This is feature flagged in master, but not currently further looked at, right?

**Comment by acinader** [comment]
I'm the wrong person to ask.  I don't use it in production.  @dplewis or @davimacedo might have a better idea of what is outstanding to make this the default.

**Comment by mtrezza** [comment]
I just realized that there is a newer thread from 2019. https://github.com/parse-community/parse-server/issues/5442 I'll ask from there.
